### PR TITLE
Support HDFS federation viewfs

### PR DESCRIPTION
**Pull request by fhoering**
## What changes are proposed in this pull request?

Support logging artifacts with Hadoop viewfs using the same fix as in Apache arrow:
https://github.com/apache/arrow/pull/5957

Link to viewfs in Hadoop:
https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/ViewFs.html

## How is this patch tested?

- tested on our Criteo Hadoop cluster having viewfs set up

### Is this a user-facing change?

- [] No. You can skip the rest of this section.
- [] Yes. Give a description of this change to be included in the release notes for MLflow users.

If you are using HDFS you can now use federation with viewfs to log artifacts.

### What component(s) does this PR affect?

- [ ] UI
- [ ] Command Line Interface
- [ ] API
- [ ] Examples
- [ ] Docs
- [ ] Tracking
- [ ] Projects
- [x ] Artifacts
- [ ] Models
- [ ] Model Registry
- [ ] Scoring
- [ ] R
- [ ] Java
- [x ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by dbczumar** [comment]
@fhoering Thanks for filing this; I'll go ahead and review it shortly. Do you have any tips for setting up and testing ViewFS?

**Comment by fhoering** [comment]
I guess this is the general problem with Hadoop and why it is declining. Nothing is easy.
https://github.com/apache/arrow/pull/5957
I

**Comment by fhoering** [comment]
I guess this is the general problem with Hadoop, nothing is easy. 

When doing https://github.com/apache/arrow/pull/5957 I had a look at the integration tests. 

They use an external docker image (seems like it is this one https://github.com/parrot-stream/docker-impala) for hdfs with a very basic hdfs config. So it would have meant adding viewfs support to this which basically means following the viewfs documentation and setting up multiple namenodes. So, as I have never done it either (as it is already set up on our cluster) and given that it was a considerable amount of work I decided not to do it.

In general, it would be nice also to have real integration tests with HDFS for mlflow. This project seems pretty nice: https://github.com/jcrist/hadoop-test-cluster, but viewfs also has to be configured as a feature. I still have this on my todo list to add hadoop integration tests to our github projects that use Hadoop. So then it would be easy to also just upstream some hdfs integration tests in mlflow.

So to sum it up. The only solution I see is setting up a hadoop cluster with docker compose and following the viewfs documentation until it works.

To come back to the fix here, basically the trick is to inject the schema to libhdfs driver. I can confirm it doesn't create a regression because pyarow integrations tests still work, it works on our cluster and in https://github.com/apache/arrow/pull/5957 there is a link to the Hadoop libhdfs code to see how the schema is handled.

**Comment by fhoering** [comment]
@dbczumar 

OK. After having asked the experts it turns out it is quite easy. You can just reuse one namenode and add the mappings to core-site.xml:
```
    <property>
        <name>fs.defaultFS</name>
        <value>viewfs://root</value>
    </property>
    <property>
        <name>fs.viewfs.mounttable.root.link./tmp</name>
        <value>hdfs://CLUSTER/tmp</value>
    </property>
    <property>
        <name>fs.viewfs.mounttable.root.link./user</name>
        <value>hdfs://CLUSTER/user</value>
    </property>
    <property>
        <name>fs.viewfs.mounttable.root.link./var</name>
        <value>hdfs://CLUSTER/var</value>
    </property>
    <property>
        <name>fs.viewfs.rename.strategy</name>
        <value>SAME_FILESYSTEM_ACROSS_MOUNTPOINT</value>
    </property>
```

**Review by dbczumar — APPROVED** [review]
LGTM @fhoering ! Sorry for the delay - I tested this out with single-node HDFS and ViewFS. I've also confirmed that HDFS URIs are always of the form `scheme://authority/path` (as discussed here: https://hadoop.apache.org/docs/r2.4.1/hadoop-project-dist/hadoop-common/FileSystemShell.html), so the URI construction on line 174 appears to be safe.

Thank you for your contribution!

**Comment by fhoering** [comment]
Thanks
